### PR TITLE
8692 notify finish fast

### PR DESCRIPTION
**Pull request by glyph**
Fix for https://twistedmatrix.com/trac/ticket/8692

### CODE REVIEW

File: src/twisted/web/http.py [hunk 1]
```diff
@@ -1784,7 +1784,7 @@ class HTTPChannel(basic.LineReceiver, policies.TimeoutMixin):
     A receiver for HTTP requests.
 
     The L{HTTPChannel} provides L{interfaces.ITransport} and
-    L{interfaces.IConsumer} to the L{Request} objects it creates. It also
+    L{interfaces.IConsumer} to the L{Request} objects it creates.  It also
```
Comment (glyph) [inline src/twisted/web/http.py:1816]: Please pardon the irrelevant whitespace changes here; this was *almost* a https://github.com/glyph/python-docstring-mode -compliant docstring, and I re-wrapped it; it's like someone went through and just deleted the double-spacing after it had been already reflowed.  (Maybe an earlier version had a bug?)

**Review by Lukasa — APPROVED** [review]
This looks reasonable to me. 

Sorry about introducing the original regression. 🙁

File: src/twisted/web/http.py [hunk 2]
```diff
@@ -2048,22 +2071,30 @@ def allContentReceived(self):
...
 
 
-    def rawDataReceived(self, data):
-        self.resetTimeout()
-
+    def dataReceived(self, data):
+        """
+        Data was received from the network.  Process it.
+        """
         # If we're currently handling a request, buffer this data. We shouldn't
```
Comment (mithrandi) [inline src/twisted/web/http.py]: This comment seems to be seriously misleading now, could you update it to describe the new behaviour? (We have _not_ necessarily paused the transport yet)

File: src/twisted/web/http.py [hunk 3]
```diff
@@ -2048,22 +2071,30 @@ def allContentReceived(self):
...
         # have received it (we've paused the transport), but let's be cautious.
```
Comment (njsmith) [inline src/twisted/web/http.py]: This comment is no longer quite true.

**Comment by glyph** [comment]
> This looks reasonable to me.

Doesn't count unless it's in the doobleydoo - https://twistedmatrix.com/trac/ticket/8692

**Review by mithrandi — APPROVED** [review]
Could you expand IRequest.notifyFinish's docstring to mention that pausing the channel from the application side will render it ineffective until the channel is resumed? While this isn't technically a new issue at all, it seems to fit with the theme of this branch.

I think this branch is ready to merge after the two issues mentioned are resolved, so I'm approving.

**Comment by glyph** [comment]
> Sorry about introducing the original regression. 🙁

No worries. HTTP/2 was a massive, risky project, and I'm very glad we undertook it.  The nature of risk is sometimes there are downsides.  I think we have generally learned a lot, and the pace of these has slowed down somewhat :-).

The regression was unfortunate, but as I noted on the trac ticket, the implementation was considerably more maintainable (and, apparently, reviewable!) than it would have been in the past.  I strongly suspect it has somewhat higher test coverage percentage these days as well.

**Comment by glyph** [comment]
Thanks @mithrandi - will do!

**Comment by glyph** [comment]
My addressing of the review feedback is complete, so if anyone comes by and sees a passing test status, feel free to merge it.

File: src/twisted/web/http.py [hunk 4]
```diff
@@ -1875,6 +1933,7 @@ class HTTPChannel(basic.LineReceiver, policies.TimeoutMixin):
     _requestProducerStreaming = None
     _waitingForTransport = False
     _abortingCall = None
+    _optimisticEagerReadSize = 0x4000
```
Comment (hawkowl) [inline src/twisted/web/http.py:1936]: how big is this

Comment (glyph) [reply src/twisted/web/http.py:1936]: 0b100000000000000

Comment (hawkowl) [reply src/twisted/web/http.py:1936]: omg glyph

Comment (glyph) [reply src/twisted/web/http.py:1936]: not sure how else to put it. ord('@') << 0o10?

Comment (hawkowl) [reply src/twisted/web/http.py:1936]: G L Y P H

Comment (glyph) [reply src/twisted/web/http.py:1936]: oh wait sorry did you want something in *decimal*?

Comment (hawkowl) [reply src/twisted/web/http.py:1936]: I may be a millennial but that doesn't mean I can read hexadecimal

Comment (glyph) [reply src/twisted/web/http.py:1936]: 50735 ^ 34351

**Comment by glyph** [comment]
hello codecov

untitled

File: src/twisted/web/http.py (hunk 4 above)
Comment (hawkowl) [reply src/twisted/web/http.py:1936]: glyph why

**Comment by njsmith** [comment]
Arguably the best value for the trigger size would just be 1 (or `0x1` or `0o1` or `0xcafe ** 0x10000 % 65537` or not written at all because you can just unconditionally call `pauseProducing` from `dataReceived`), i.e., all non-pipelining clients get good behavior and all pipelining clients get shut down.

**Comment by glyph** [comment]
> Arguably the best value for the trigger size would just be 1 (or 0x1 or 0o1 or 0xcafe ** 0x10000 % 65537 or not written at all because you can just unconditionally call pauseProducing from dataReceived), i.e., all non-pipelining clients get good behavior and all pipelining clients get shut down.

We should probably open a new ticket in Trac to discuss.

**Comment by njsmith** [comment]
[ugh fine](https://twistedmatrix.com/trac/ticket/9203)

**Pull request merged**